src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,93 @@
+// This is an open source non-commercial project. Dear PVS-Studio, please check
+// it. PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
+
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+
+#include "nvim/api/private/helpers.h"
+#include "nvim/api/private/defs.h"
+#include "nvim/lua/executor.h"
+#include "nvim/vim.h"
+#include "nvim/version.h"
+
+#ifdef WIN32
+#include <Winuser.h>
+#endif
+
+#ifdef __APPLE__
+#define Boolean Boolean_I_Dont_Care
+#include <IOKit/IOKitLib.h>
+#include <IOKit/IOReturn.h>
+#include <IOKit/hidsystem/IOHIDLib.h>
+#include <IOKit/hidsystem/IOHIDParameter.h>
+#include <CoreFoundation/CoreFoundation.h>
+#undef Boolean
+#endif
+
+/// @return -1 if error encountered, 1 if capslock active otherwise 0 if not.
+bool capslock_status(const char **const err )
+{
+  *err = NULL;
+#if defined(__linux__)
+  String cmd = cstr_to_string("return (function()"
+                              "local handle = io.popen('xset q|grep Caps')"
+                              "local data = handle:read('*all')"
+                              "handle:close()"
+                              "return data end)()");
```

Comment:
This will be all on one line which isn't valid Lua syntax. You would need newlines:

```
"return (function()\n"
"local handle = io.popen('xset q|grep Caps')\n"
...
```

However I would suggest putting the Lua function definition into `src/nvim/lua/vim.lua`. See `_os_proc_info` for example. Remember to add it to the "exported" list at the end of `src/nvim/lua/vim.lua`.

Reply:
OK, cool, this is better, I had hard time in finding this, I had to ask from lua developers on their IRC, OK I will use your approach :smile:

Reply:
@nimitbhardwaj You could also use the `_system()` Lua function in that same file (again see `_os_proc_info` and `_os_proc_children`, they both use `_system()`) instead of popen.